finality-aleph/src/data_io.rs

```diff
@@ -203,138 +448,535 @@ where
+                self.remove_id_from_pending_proposal(&proposal, id);
             }
+            true
         } else {
-            self.add_pending_message(message, requirements);
+            error!(target:"afa", "Tried to prune a message but there is none pending.");
+            false
         }
     }
 
-    fn push_messages(&mut self, block_data: AlephDataFor<B>) {
-        if let Some(MissingBlockInfo {
-            messages: ids,
-            first_occurence: _,
-        }) = self.missing_blocks.remove(&block_data)
+    // Checks if we have exceeded the maximum number of pending messages or proposals.
+    // If so, we prune messages until the limits are satisfied again.
+    fn prune_pending_messages(&mut self) {
+        while self.pending_messages.len() > self.config.max_messages_pending
+            || self.pending_proposals.len() > self.config.max_proposals_pending
         {
-            for message_id in ids.iter() {
-                *self
-                    .message_requirements
-                    .get_mut(message_id)
-                    .expect("there are some requirements") -= 1;
-                if self.message_requirements[message_id] == 0 {
-                    let message = self
-                        .pending_messages
-                        .remove(message_id)
-                        .expect("there is a pending message");
-                    if let Err(e) = self.messages_for_aleph.unbounded_send(message) {
-                        debug!(target: "afa", "Unable to send a ready message from DataStore {}", e);
-                    }
-                    self.message_requirements.remove(message_id);
-                }
+            if !self.prune_single_message() {
+                error!(target:"afa", "Message pruning in DataStore failed. Moving on.");
+                break;
             }
         }
     }
 
-    fn add_block(&mut self, block_data: AlephDataFor<B>) {
-        debug!(target: "afa", "Adding block {:?} to Data Store", block_data);
-        self.available_blocks.put(block_data, ());
-        self.push_messages(block_data);
+    fn prune_triggers(&mut self) {
+        if self.num_triggers_registered_since_last_pruning > MAX_PROPOSALS_PENDING {
+            // Prune all the data that is not pending anymore and all the events that
+            // have an empty list if triggers.
+            let pending_proposals = &self.pending_proposals;
+            self.event_triggers.retain(|_event, proposal_set| {
+                proposal_set.retain(|proposal| pending_proposals.contains_key(proposal));
+                !proposal_set.is_empty()
+            });
+            self.num_triggers_registered_since_last_pruning = 0;
+            //TODO make sure we increment this
+        }
+    }
+
+    fn on_message_received(&mut self, message: Message) {
+        let included_data = message.included_data();
+        if included_data.iter().any(|d| !d.is_correct()) {
```

Comment:
It's a single message that might contain multiple pieces of data (e.g. a set of units with differing contents). At this level of abstraction it's impossible to separate the pieces (it might be impossible in general) and we wouldn't get such a message from an honest node anyway.

Reply:
okay, but if so, then can we somehow mark/announce dishonesty of the sender?

Reply:
That's annoyingly complicated at this moment – the Substrate reporting system does not do this the way we would like, and we don't have one of our own that would be good enough.

Reply:
The thing we might (and should) at some point do is disconnect from peers sending us 1) incorrect data, 2) too many messages, too quickly. But this is a non-trivial task, and requires many changes that touch upon the network layer. So unfortunately, future work.